addons/website_theme_install/models/ir_module_module.py

```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
-        menu_todel = menus.with_context(active_test=False).mapped('copy_ids').filtered(lambda m: m.website_id == website)
-        menu_todel.unlink()
-
-        pages = self._get_module_data('theme.website.page')
-        pages_todel = pages.with_context(active_test=False).mapped('copy_ids').filtered(lambda p: p.website_id == website)
-        pages_todel.unlink()
-
-        attachs = self._get_module_data('theme.ir.attachment')
-        attachs_todel = attachs.with_context(active_test=False).mapped('copy_ids').filtered(lambda a: a.website_id == website)
-        # double check - Some records that can be orphans or no more in the template view.
-        attachs_todel2 = self.env['ir.attachment'].with_context(active_test=False).search([('key', '=like', self.name + '_%'), ('website_id', '=', website.id)])
-        (attachs_todel + attachs_todel2).unlink()
-
-        views = self._get_module_data('theme.ir.ui.view')
-        views_todel = views.with_context(active_test=False).mapped('copy_ids').filtered(lambda v: v.website_id == website)
-        # double check - Some records that can be orphans or no more in the template view.
-        views_todel2 = self.env['ir.ui.view'].with_context(active_test=False).search([('key', '=like', self.name + '_%'), ('website_id', '=', website.id)])
-        (views_todel + views_todel2).unlink()
+
+        model = self.env[model_name]
+
+        if model_name in ('website.page', 'website.menu'):
+            return model
+
+        return model.with_context(active_test=False).search([
+            ('key', '=like', self.name + '.%'),
+            ('website_id', '=', website.id),
+            ('theme_template_id', '=', False),
+        ])
 
     @api.multi
-    def _remove_theme_on_website(self, website):
+    def _theme_cleanup(self, model_name, website):
+        """Remove orphan models of the current theme for the website ``website``."""
         self.ensure_one()
-        installed_deps = self + website.theme_id.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
-        for mod in installed_deps:
-            mod._unload_one_theme_module(website)
+        orphans = self._theme_find_orphans(model_name, website)
+        orphans.unlink()
+        return orphans
 
     @api.multi
-    def _add_theme_on_website(self, website):
+    def _theme_get_upstream(self):
+        """Return installed upstream themes."""
         self.ensure_one()
-        mods_to_load = reversed(self + self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_')))
-        for mod in mods_to_load:
-            _logger.info('Load theme %s for website %s from template.' % (mod.name, website.id))
-            mod._load_one_theme_module(website, with_update=False)
-            self.env['theme.utils']._post_copy(mod)
+        return self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
```

Comment:
Reversed is needed only on "_remove_theme" to remove them on the opposite order they were added.

Reversed is not needed anymore in the other cases because the different "_theme_get_stream_themes" methods return them in the correct order on purpose.

As for the upgrade is triggered automatically in the write, so it will always be in the correct order because that's how modules are upgraded in general.